contributors/design-proposals/defer-containers.md

```diff
@@ -0,0 +1,248 @@
+# Defer Containers
+
+dhilip.kumar.s@huawei.com
+
+March 2017
+
+## Prerequisite
+This is a continuation of [initContainers](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/container-init.md)
+
+## Motivation
+Since the introduction to Statefulset and Daemonset PODs that consist of stateful containers are now fully supported by
+Kubernetes.  Such stateful pods benefit from constructor and destructor semantics. InitContainers provide constructor
+semantics. The proposed DeferContainers will provide destructor semantics for such stateful PODs
+
+## Abstract
+This introduces the concept of ‘deferContainers’, inspired by golang’s keyword ‘defer’. This will allow one to define a
+set of Containers to be executed at the time of POD termination. The defined containers will be executed sequentially and
+in the specified order. It will bring destructor() capability to Pods.
+
+## Use Cases
+
+### Cleanup:
```

Comment:
How are we going to handle situation:
- you want to delete pod because it's broken or whatever
- it has deferContainer that will hung because of "some reasons"

How am I supposed to *not* wait for deferContainer? Are we going to add `kubectl delete .... --ignore-defer-containers`?